README.md
# weighted_rand

[![weighted_rand](https://github.com/ichi-h/weighted_rand/actions/workflows/weighted_rand.yml/badge.svg)](https://github.com/ichi-h/weighted_rand/actions/workflows/weighted_rand.yml)
[![Crates.io](https://img.shields.io/crates/v/weighted_rand)](https://crates.io/crates/weighted_rand)
[![docs.rs](https://img.shields.io/docsrs/weighted_rand)](https://docs.rs/weighted_rand)
[![Crates.io](https://img.shields.io/crates/l/weighted_rand)](LICENSE-APACHE)

A weighted random sampling crate using Walker's Alias Method.

Walker's Alias Method (WAM) is one method for performing weighted random sampling.  
WAM weights each index of a array by giving two pieces of information: an alias to a different index and a probability to decide whether to jump to that index.

WAM is a very fast algorithm, and its computational complexity of the search is O(1).  
The difference in complexity between WAM and the Cumulative Sum Method is as follows.

|       Algorithm       | Building table |  Search  |
| :-------------------: | :------------: | :------: |
| Walker's Alias Method |      O(N)      |   O(1)   |
| Cumulative Sum Method |      O(N)      | O(log N) |

The API documentation is [here](https://docs.rs/weighted_rand).

## Usage

Add this to your Cargo.toml:

```toml
[dependencies]
weighted_rand = "0.4"
```

## Example

```rust
use weighted_rand::builder::WalkerTableBuilder;

fn main() {
    let fruit = ["Apple", "Banana", "Orange", "Peach"];

    // Define the weights for each index corresponding
    // to the above list.
    // In the following case, the ratio of each weight
    // is "2 : 1 : 7 : 0", and the output probabilities
    // for each index are 0.2, 0.1, 0.7 and 0.
    let index_weights = [2, 1, 7, 0];

    let builder = WalkerTableBuilder::new(&index_weights);
    let wa_table = builder.build();

    for i in (0..10).map(|_| wa_table.next()) {
        println!("{}", fruit[i]);
    }
}
```

Also, `index_weiaghts` supports `&[f32]`, like:

```rust
use rand;
use weighted_rand::builder::*;

fn main() {
    // Coin with a 5% higher probability of heads than tails
    let cheating_coin = ["Heads!", "Tails!"];
    let index_weights = [0.55, 0.45];

    let builder = WalkerTableBuilder::new(&index_weights);
    let wa_table = builder.build();

    // If you want to process something in a large number of
    // loops, we recommend using the next_rng method with an
    // external ThreadRng instance.
    let mut result = [""; 10000];
    let mut rng = rand::thread_rng();
    for r in &mut result {
        let j = wa_table.next_rng(&mut rng);
        *r = cheating_coin[j];
    }

    // println!("{:?}", result);
}
```

## License

Licensed under either of

- Apache License, Version 2.0
  ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
- MIT license
  ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

## Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be
dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
# Change Log

## 0.4.2 - 2023-10-2

### Changes

- [#5](https://github.com/ichi-h/weighted_rand/pull/5) Thanks [@DanArmor](https://github.com/DanArmor)! - fix: return 1 for gcd_for_slice if all elements are 0.0

## 0.4.1 - 2023-7-3

### Changes

- [#3](https://github.com/ichi-h/weighted_rand/pull/3) Thanks [@ramon54321](https://github.com/ramon54321)! - Allow any Rng implementing type in next_rng
- [#4](https://github.com/ichi-h/weighted_rand/pull/4) Thanks [@Braymatter](https://github.com/Braymatter)! - Added Clone and Default derives to WalkerTable

## 0.3.2 - 2021-11-11

### Changes

- Update basic usage example.
- Change unnecessary `&Vec<T>` to `&[T]`

## 0.3.1 - 2021-11-9

### Changes

- Implementation of `WalkerTable::next_rng()` method.
- Add benchmark test.
- Performance improvement.
  - Fix type of aliases from `Vec<u32>` to `Vec<usize>`. ([b362a3f](https://github.com/ichi-h/weighted_rand/commit/b362a3f11ba1505fd733ed208562c2f91e5b1f2a))
  - Remove unnecessary else statement. ([87909d7](https://github.com/ichi-h/weighted_rand/commit/87909d744339eb4307b1192aaaac7807c13fadaf))
  - Change threshold (`Vec<u32>`) to probabilities (`Vec<f32>`). ([fccb34d](https://github.com/ichi-h/weighted_rand/commit/fccb34d60e3a13c7fefaca09fd594819f15a5ff8))

## 0.3.0 - 2021-11-9

### Changes

- Implementation of `Vec<f32>` type for `index_weights`. [#2](https://github.com/ichi-h/weighted_rand/pull/2)

## 0.2.0 - 2021-11-8

### Changes

- `WalkerTableBuilder::build()` returns `WalkerTable` instead of `Result<WalkerTable>`.

## 0.1.0 - 2021-11-8

- First release.


src/util.rs
pub mod math {
    fn gcd(a: u32, b: u32) -> u32 {
        let (a, b) = if a < b { (b, a) } else { (a, b) };

        if a % b == 0 {
            b
        } else {
            gcd(b, a % b)
        }
    }

    pub fn gcd_for_slice(slice: &[u32]) -> u32 {
        if slice.is_empty() {
            return 0;
        }

        let mut iter = slice.iter().skip_while(|x| x == &&0);
        let first = match iter.next() {
            Some(v) => *v,
            None => return 1
        };

        let gcd = iter.fold(
            first,
            |acc, cur| {
                if *cur == 0 {
                    acc
                } else {
                    gcd(*cur, acc)
                }
            },
        );
        gcd
    }
}

#[cfg(test)]
mod math_test {
    use crate::util::math::gcd_for_slice;

    #[test]
    fn gcd_pattern_1() {
        let actual = gcd_for_slice(&[4, 20, 32]);
        assert_eq!(actual, 4)
    }

    #[test]
    fn gcd_pattern_2() {
        let actual = gcd_for_slice(&[77, 9, 25]);
        assert_eq!(actual, 1)
    }

    #[test]
    fn gcd_pattern_3() {
        let actual = gcd_for_slice(&[11, 0, 22]);
        assert_eq!(actual, 11)
    }

    #[test]
    fn gcd_pattern_4() {
        let actual = gcd_for_slice(&[]);
        assert_eq!(actual, 0)
    }
}


src/table.rs
//! Weighted random index generator by Walker's Alias Method.

use rand::prelude::*;
use serde::{Deserialize, Serialize};

/// Table of aliases and probabilities
///
/// In Walker's Alias Method, weighted random sampling is performed by the
/// following operations.
///
/// 1. Get an index "i" randomly.
/// 2. Get a random value "r" between 0 and 1
/// 3. If "r" exceeds the value of `probs[i]`, "i" is output as-is. If it
///    does not, the value of `aliases[i]` (which means the alias to another
///    index) is output.
///
/// The more likely a particular index is to be output, the more of the value
/// will be included in `aliases`.
#[derive(Serialize, Deserialize, Debug, PartialEq, Clone, Default)]
pub struct WalkerTable {
    /// Alias to another index
    aliases: Vec<usize>,

    /// Probability for whether to output the index attached to `aliases`.
    probs: Vec<f32>,
}

impl WalkerTable {
    /// Creates a new instance of [`WalkerTable`].
    pub fn new(aliases: Vec<usize>, probs: Vec<f32>) -> WalkerTable {
        WalkerTable {
            aliases: aliases,
            probs: probs,
        }
    }

    /// Returns an index at random.
    pub fn next(&self) -> usize {
        let mut rng = rand::thread_rng();
        self.next_rng(&mut rng)
    }

    /// Returns an index at random using an external RNG which implements Rng.
    pub fn next_rng(&self, rng: &mut impl Rng) -> usize {
        let i = rng.gen_range(0..self.probs.len());
        let r = rng.gen::<f32>();
        if r < self.probs[i] {
            return self.aliases[i];
        }
        i
    }
}

#[cfg(test)]
mod table_test {
    use crate::builder::*;
    use rand;

    const N: usize = 100_000;

    fn count<T: PartialEq>(slice: &[T], target: T) -> f32 {
        slice
            .iter()
            .fold(0.0, |acc, cur| if *cur == target { acc + 1.0 } else { acc })
    }

    #[test]
    fn unweighted_random_sampling() {
        let index_weights = [0; 4];
        let builder = WalkerTableBuilder::new(&index_weights);
        let wa_table = builder.build();

        let mut rng = rand::thread_rng();

        let idxs = (0..N)
            .map(|_| wa_table.next_rng(&mut rng))
            .collect::<Vec<usize>>();

        let i_0 = count(&idxs, 0);
        let i_1 = count(&idxs, 1);
        let i_2 = count(&idxs, 2);
        let i_3 = count(&idxs, 3);

        const EXPT: f32 = N as f32 * 0.25;
        assert!(
            (EXPT * 0.95 < i_0 && i_0 < EXPT * 1.05)
                && (EXPT * 0.95 < i_1 && i_1 < EXPT * 1.05)
                && (EXPT * 0.95 < i_2 && i_2 < EXPT * 1.05)
                && (EXPT * 0.95 < i_3 && i_3 < EXPT * 1.05)
        )
    }

    #[test]
    fn weighted_random_sampling() {
        let index_weights = [2, 1, 7, 0];
        let builder = WalkerTableBuilder::new(&index_weights);
        let wa_table = builder.build();

        let idxs = (0..N).map(|_| wa_table.next()).collect::<Vec<usize>>();

        let i_0 = count(&idxs, 0);
        let i_1 = count(&idxs, 1);
        let i_2 = count(&idxs, 2);
        let i_3 = count(&idxs, 3);

        const EXPT: [f32; 4] = [N as f32 * 0.2, N as f32 * 0.1, N as f32 * 0.7, 0.0];
        assert!(
            (EXPT[0] * 0.95 < i_0 && i_0 < EXPT[0] * 1.05)
                && (EXPT[1] * 0.95 < i_1 && i_1 < EXPT[1] * 1.05)
                && (EXPT[2] * 0.95 < i_2 && i_2 < EXPT[2] * 1.05)
                && (EXPT[3] == i_3)
        )
    }
}


src/builder.rs
//! Builds a [`WalkerTable`] instance.

use crate::table::WalkerTable;
use crate::util::math::gcd_for_slice;

pub trait NewBuilder<T> {
    /// Creates a new instance of [`WalkerTableBuilder`] from
    /// [`&[u32]`] or [`&[f32]`].
    ///
    /// Values less than 0 will be calculated as 0.
    fn new(index_weights: &[T]) -> WalkerTableBuilder;
}

/// Builder of [`WalkerTable`]
///
/// ## Example
///
/// ```rust
/// use weighted_rand::builder::*;
///
/// fn main() {
///     let index_weights = [1, 2, 3, 4];
///     let builder = WalkerTableBuilder::new(&index_weights);
///     let wa_table = builder.build();
/// }
/// ```
///
/// Also, `index_weiaghts` supports [`&[f32]`], like `[0.1, 0.2, 0.3, 0.4]`
///
/// ## About `index_weights`
///
/// `index_weights` is the weights of the output indexes.
///
/// The larger the value, the more likely the corresponding index will be
/// output.
///
/// For example, if this value is `[2, 1, 7, 0]`, the output probabilities
/// for each index are 0.2, 0.1, 0.7 and 0. If a weight value is 0, the
/// corresponding index will not be output. In other words, the index 3 will
/// not be output in the this cases.
pub struct WalkerTableBuilder {
    /// Weights of the output indexes.
    index_weights: Vec<u32>,
}

impl NewBuilder<u32> for WalkerTableBuilder {
    fn new(index_weights: &[u32]) -> WalkerTableBuilder {
        let table_len = index_weights.len() as u32;

        // Process that the mean of index_weights does not become a float value
        let ws = index_weights
            .iter()
            .map(|w| w * table_len)
            .collect::<Vec<u32>>();

        WalkerTableBuilder { index_weights: ws }
    }
}

impl NewBuilder<f32> for WalkerTableBuilder {
    fn new(index_weights: &[f32]) -> WalkerTableBuilder {
        let ws = index_weights
            .iter()
            .map(|w| (w * 10000.0).round() as u32)
            .collect::<Vec<u32>>();

        let gcd = gcd_for_slice(&ws);
        let ws = ws.iter().map(|w| w / gcd).collect::<Vec<u32>>();

        WalkerTableBuilder::new(&ws)
    }
}

impl WalkerTableBuilder {
    /// Builds a new instance of [`WalkerTable`].
    pub fn build(&self) -> WalkerTable {
        let table_len = self.index_weights.len();

        if self.sum() == 0 {
            // Returns WalkerTable that performs unweighted random sampling.
            return WalkerTable::new(vec![0; table_len], vec![0.0; table_len]);
        }

        let (aliases, probs) = self.calc_table();

        WalkerTable::new(aliases, probs)
    }

    /// Inverses given weights
    pub fn inverse(self) -> WalkerTableBuilder {
        let min_value = match self.index_weights.iter().min() {
            Some(v) => *v,
            None => 0,
        };
        let max_value = match self.index_weights.iter().max() {
            Some(v) => *v,
            None => 0,
        };
        Self {
            index_weights: self
                .index_weights
                .into_iter()
                .map(|x| {
                    if x == max_value {
                        min_value
                    } else if x == min_value {
                        max_value
                    } else {
                        max_value - x
                    }
                })
                .collect(),
        }
    }

    /// Calculates the sum of `index_weights`.
    fn sum(&self) -> u32 {
        self.index_weights.iter().fold(0, |acc, cur| acc + cur)
    }

    /// Calculates the mean of `index_weights`.
    fn mean(&self) -> u32 {
        self.sum() / self.index_weights.len() as u32
    }

    /// Returns the tables of aliases and probabilities.
    fn calc_table(&self) -> (Vec<usize>, Vec<f32>) {
        let table_len = self.index_weights.len();
        let (mut below_vec, mut above_vec) = self.separate_weight();
        let mean = self.mean();

        let mut aliases = vec![0; table_len];
        let mut probs = vec![0.0; table_len];
        loop {
            match below_vec.pop() {
                Some(below) => {
                    if let Some(above) = above_vec.pop() {
                        let diff = mean - below.1;
                        aliases[below.0] = above.0 as usize;
                        probs[below.0] = diff as f32 / mean as f32;
                        if above.1 - diff <= mean {
                            below_vec.push((above.0, above.1 - diff));
                        } else {
                            above_vec.push((above.0, above.1 - diff));
                        }
                    } else {
                        aliases[below.0] = below.0 as usize;
                        probs[below.0] = below.1 as f32 / mean as f32;
                    }
                }
                None => break,
            }
        }

        (aliases, probs)
    }

    /// Divide the values of `index_weights` based on the mean of them.
    ///
    /// The tail value is a weight and head is its index.
    fn separate_weight(&self) -> (Vec<(usize, u32)>, Vec<(usize, u32)>) {
        let mut below_vec = Vec::with_capacity(self.index_weights.len());
        let mut above_vec = Vec::with_capacity(self.index_weights.len());
        for (i, w) in self.index_weights.iter().enumerate() {
            if *w <= self.mean() {
                below_vec.push((i, *w));
            } else {
                above_vec.push((i, *w));
            }
        }
        (below_vec, above_vec)
    }
}

#[cfg(test)]
mod builder_test {
    use crate::builder::*;
    use crate::table::WalkerTable;

    #[test]
    fn make_table_from_u32() {
        let index_weights = [2, 7, 9, 2, 4, 8, 1, 3, 6, 5];
        let builder = WalkerTableBuilder::new(&index_weights);
        let w_table = builder.build();

        let expected = WalkerTable::new(
            vec![2, 1, 1, 2, 2, 2, 5, 9, 5, 8],
            vec![
                0.574468085106383,
                1.0,
                0.48936170212766,
                0.574468085106383,
                0.148936170212766,
                0.106382978723404,
                0.787234042553192,
                0.361702127659574,
                0.0212765957446809,
                0.297872340425532,
            ],
        );

        assert_eq!(w_table, expected)
    }

    #[test]
    fn make_table_from_f32() {
        let index_weights = [0.1, 0.2, 0.3, -0.4];
        let builder = WalkerTableBuilder::new(&index_weights);
        let w_table = builder.build();

        let expected = WalkerTable::new(vec![1, 1, 1, 2], vec![0.333333333333333, 1.0, 0.0, 1.0]);

        assert_eq!(w_table, expected)
    }

    #[test]
    fn when_sum_is_zero() {
        let index_weights = [0; 5];
        let builder = WalkerTableBuilder::new(&index_weights);
        let w_table = builder.build();

        let expected = WalkerTable::new(vec![0; 5], vec![0.0; 5]);

        assert_eq!(w_table, expected)
    }
}


benches/benchmark.rs
use rand::prelude::*;
use weighted_rand::builder::*;

use criterion::Criterion;
use criterion::{criterion_group, criterion_main};

const WEIGHTS: [u32; 1000] = [1; 1000];

fn bench_constructor(c: &mut Criterion) {
    c.bench_function("constructor", |b| {
        b.iter(|| WalkerTableBuilder::new(&WEIGHTS))
    });
}

fn bench_generate_by_wam_next(c: &mut Criterion) {
    let builder = WalkerTableBuilder::new(&WEIGHTS);
    let table = builder.build();

    let mut result = [0; 100_000];

    c.bench_function("generate_by_wam_next", |b| {
        b.iter(|| {
            for r in &mut result {
                *r = table.next();
            }
        })
    });
}

fn bench_generate_by_wam_next_rng(c: &mut Criterion) {
    let builder = WalkerTableBuilder::new(&WEIGHTS);
    let table = builder.build();

    let mut rng = rand::thread_rng();

    let mut result = [0; 100_000];

    c.bench_function("generate_by_wam_next_rng", |b| {
        b.iter(|| {
            for r in &mut result {
                *r = table.next_rng(&mut rng);
            }
        })
    });
}

fn bench_generate_by_csm(c: &mut Criterion) {
    let probs = (0..1000)
        .map(|v| (v + 1) as f32 / 1000.0)
        .collect::<Vec<f32>>()
        .to_vec();

    let csm = CSM { probs: &probs };
    let mut rng = rand::thread_rng();
    let mut result = [0; 100_000];

    c.bench_function("generate_by_csm", |b| {
        b.iter(|| {
            for r in &mut result {
                *r = csm.next(&mut rng);
            }
        })
    });
}

criterion_group!(
    benches,
    bench_constructor,
    bench_generate_by_wam_next,
    bench_generate_by_wam_next_rng,
    bench_generate_by_csm
);
criterion_main!(benches);

// ========================================================= //

// Weighted random sampling using Cumulative Sum Method

struct CSM<'a> {
    probs: &'a [f32],
}

impl CSM<'_> {
    fn next(&self, rng: &mut ThreadRng) -> usize {
        let r = rng.gen::<f32>();
        for (i, p) in self.probs.iter().enumerate() {
            if r <= *p {
                return i;
            }
        }

        self.probs.len() - 1
    }
}


src/lib.rs
//! # weighted_rand
//!
//! A weighted random sampling crate using Walker's Alias Method.
//!
//! Walker's Alias Method (WAM) is one method for performing weighted
//! random sampling. WAM weights each index of a array by giving two
//! pieces of information: an alias to a different index and a
//! probability to decide whether to jump to that index.
//!
//! WAM is a very fast algorithm, and its computational complexity of
//! the search is O(1). The difference in complexity between WAM and
//! the Cumulative Sum Method is as follows.
//!
//! |       Algorithm       | Building table |  Search  |
//! | :-------------------: | :------------: | :------: |
//! | Walker's Alias Method |      O(N)      |   O(1)   |
//! | Cumulative Sum Method |      O(N)      | O(log N) |
//!
//! ## Example
//!
//! ```rust
//! use weighted_rand::builder::*;
//!
//! fn main() {
//!     let fruit = ["Apple", "Banana", "Orange", "Peach"];
//!
//!     // Define the weights for each index corresponding
//!     // to the above list.
//!     // In the following case, the ratio of each weight
//!     // is "2 : 1 : 7 : 0", and the output probabilities
//!     // for each index are 0.2, 0.1, 0.7 and 0.
//!     let index_weights = [2, 1, 7, 0];
//!
//!     let builder = WalkerTableBuilder::new(&index_weights);
//!     let wa_table = builder.build();
//!
//!     for i in (0..10).map(|_| wa_table.next()) {
//!         println!("{}", fruit[i]);
//!     }
//! }
//! ```
//!
//! Also, `index_weiaghts` supports [`&[f32]`], like:
//!
//! ```rust
//! use rand;
//! use weighted_rand::builder::*;
//!
//! fn main() {
//!     // Coin with a 5% higher probability of heads than tails
//!     let cheating_coin = ["Heads!", "Tails!"];
//!     let index_weights = [0.55, 0.45];
//!
//!     let builder = WalkerTableBuilder::new(&index_weights);
//!     let wa_table = builder.build();
//!
//!     // If you want to process something in a large number of
//!     // loops, we recommend using the next_rng method with an
//!     // external ThreadRng instance.
//!     let mut result = [""; 10000];
//!     let mut rng = rand::thread_rng();
//!     for r in &mut result {
//!         let j = wa_table.next_rng(&mut rng);
//!         *r = cheating_coin[j];
//!     }
//!
//!     // println!("{:?}", result);
//! }
//! ```
//!

pub mod builder;
pub mod table;

mod util;
